File: advent-of-code/2025/day-5-cafeteria/normal/n-d05-12.kt
fun main() {
    val input = generateSequence(::readLine).joinToString("\n").split("\n\n")
    val ranges = input[0].lines().map { val (a, b) = it.split("-").map(String::toLong); a to b }. sortedBy { it.first }. toMutableList()
    val ids = input[1].trim().lines().map { it.toLong() }
    
    val merged = mutableListOf<Pair<Long, Long>>()
    for ((lo, hi) in ranges) {
        if (merged.isNotEmpty() && lo <= merged.last().second + 1)
            merged[merged.lastIndex] = merged.last().first to maxOf(merged.last().second, hi)
        else merged.add(lo to hi)
    }
    
    val p1 = ids.count { i -> merged.any { (lo, hi) -> i in lo..hi } }
    val p2 = merged.sumOf { (lo, hi) -> hi - lo + 1 }
    println(p1)
    println(p2)
}

File: advent-of-code/2025/day-5-cafeteria/normal/n-d05-12.rs
use std::io::{self, Read};

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let parts: Vec<&str> = input.trim().split("\n\n").collect();
    let mut ranges:  Vec<(i64, i64)> = parts[0].lines().map(|l| {
        let p: Vec<i64> = l.split('-').map(|x| x.parse().unwrap()).collect();
        (p[0], p[1])
    }).collect();
    ranges.sort();
    
    let mut merged:  Vec<(i64, i64)> = vec![];
    for (lo, hi) in ranges {
        if ! merged.is_empty() && lo <= merged.last().unwrap().1 + 1 {
            let last = merged.last_mut().unwrap();
            last.1 = last.1.max(hi);
        } else {
            merged.push((lo, hi));
        }
    }
    
    let ids: Vec<i64> = parts[1].trim().lines().map(|l| l.parse().unwrap()).collect();
    let p1 = ids.iter().filter(|&&i| merged.iter().any(|&(lo, hi)| lo <= i && i <= hi)).count();
    let p2: i64 = merged.iter().map(|(lo, hi)| hi - lo + 1).sum();
    println!("{}\n{}", p1, p2);
}
